microovn/ovn/ovsdb/ovsdb.go

```diff
@@ -26,3 +41,241 @@ func ExpectedOvsdbSchemaVersion(s *state.State, dbSpec *ovnCmd.OvsdbSpec) (strin
+		return fmt.Errorf("failed create DB specification: '%s'", err.Error())
+	}
+
+	if !dbSpec.IsCentral {
+		return fmt.Errorf("MicroOVN handles upgrades only for Northbound and Southbound clustered DBs")
+	}
+
+	centralActive, err := node.HasServiceActive(s, "central")
+	if err != nil {
+		return fmt.Errorf("failed to query local services: %s", err)
+	}
+
+	if !centralActive {
+		logger.Infof(
+			"OVN %s database won't be upgraded from this node because 'central' service is not active",
+			dbSpec.FriendlyName,
+		)
+		return nil
+	}
+
+	backOffMs := 100
+	var dbStatus schemaStatus
+
+	// Start of the loop that coordinates with other cluster members the OVSDB cluster schema upgrade.
+	for {
+		// Get current schema status
+		dbStatus, err = getLiveSchemaStatus(s, dbSpec)
+		if err != nil {
+			logger.Warnf(
+				"Error checking if OVN %s cluster needs upgrade: '%s' (retrying in %d ms)",
+				dbSpec.FriendlyName,
+				err,
+				backOffMs,
+			)
+		} else {
+			if !dbStatus.UpgradeRequired {
+				// If upgrade is not required, break out of the loop
+				logger.Infof("OVN %s DB is at expected version. No upgrade needed", dbSpec.FriendlyName)
+				break
+			} else {
+				logger.Infof("OVN %s DB schema needs upgrade.", dbSpec.FriendlyName)
+			}
+
+			// Check whether we are the designated node for triggering the schema upgrade
+			upgradeLeader, err := isNodeUpgradeLeader(s)
+			if err != nil {
+				logger.Warnf(
+					"Failed to determine if this host is OVN %s upgrade leader: '%s' (retrying in %d ms)",
+					dbSpec.FriendlyName,
+					err,
+					backOffMs,
+				)
+			} else if upgradeLeader {
+
+				// Leader verifies whether the cluster is ready for schema upgrade
+				upgradeReady, err := isClusterUpgradeReady(s, dbSpec, dbStatus.TargetVersion)
+				if err != nil {
+					logger.Warnf(
+						"OVN %s DB upgrade readycheck failed: '%s' (retrying in %d ms)",
+						dbSpec.FriendlyName,
+						err,
+						backOffMs,
+					)
+				} else if upgradeReady {
+
+					// If the cluster is ready, an upgrade is triggered. Otherwise, the loop continues.
+					logger.Infof("Triggering OVN %s schema upgrade.", dbSpec.FriendlyName)
+					_, err = ovnCmd.OvsdbClient(
+						s,
+						dbSpec,
+						10,
+						60,
+						"convert",
+						dbSpec.SocketURL,
+						fmt.Sprintf(dbSpec.Schema),
+					)
+					return err
+				}
+			} else {
+				// Nodes that are not designated to trigger the upgrade, continue looping. This ensures that in case
+				// that cluster has to be altered during the upgrade, e.g. if the original leader has to be replaced,
+				// a new member will become a designated leader, and it will trigger the upgrade.
+				logger.Infof(
+					"This host is not a designated to upgrade OVN %s DB schema. Rechecking schema status in %d",
+					dbSpec.FriendlyName,
+					backOffMs,
+				)
+			}
+		}
+
+		time.Sleep(time.Duration(backOffMs) * time.Millisecond)
+		if backOffMs < 2000 {
+			backOffMs *= 2
+		}
+
+		if backOffMs > 2000 {
+			backOffMs = 2000
```

Comment:
nit: see comment above about using constant and multipliers instead of bare numbers for this algorithm.